### PR TITLE
Allow non-Computers to have Software

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | -

Specifically to add proper support for inventory of smart phones, or VoIP phones that have apps (such as Grandstream phones which are Android based). Follows #6090 

Needed for Jamf plugin to sync iPhone software inventory cconard96/jamf#31.

### CODE REVIEW

File: inc/item_softwarelicense.class.php [hunk 1]
```diff
@@ -0,0 +1,987 @@
...
+         ];
+         if ($DB->fieldExists($itemtable, 'serial')) {
+            $query['SELECT'][] = $itemtable.'.serial';
+         } else {
+            $query['SELECT'][] = "'' AS ".$itemtable.".serial";
+         }
+         if ($DB->fieldExists($itemtable, 'otherserial')) {
+            $query['SELECT'][] = $itemtable.'.otherserial';
+         } else {
+            $query['SELECT'][] = "'' AS ".$itemtable.".otherserial";
```
Comment (cedric-anne) [inline inc/item_softwarelicense.class.php]:
```suggestion
            $query['SELECT'][] = new QueryExpression($DB->quoteValue('') . " AS " . $DB->quoteName($itemtable.".otherserial"));
```
also for other uses of `value as field` in this PR.

**Comment by cconard96** [comment]
> I quickly read this PR, but still not test it, I will do it when automatic test will pass.
> 
> In addition to remarks in code review, I noticed that `tools/generate_bigdump.function.php` should be updated to use new classes.

I updated the class names and queries in this tool, but I do not know enough about it to add new itemtypes to it now that Phones can also have software.

**Comment by cconard96** [comment]
Also, tests are passing. I cannot see the functional test output from the last push, but the latest push passes. Since the previous push was done on a weekend, it was most likely caused by #6257.

File: install/update_94_95.php [hunk 2]
```diff
@@ -486,6 +485,68 @@ function update94to95() {
...
+   );
+   $migration->addKey('glpi_items_softwarelicenses', 'itemtype');
+   $migration->dropKey('glpi_items_softwarelicenses', 'computers_id');
+   $migration->addKey('glpi_items_softwarelicenses', 'items_id', 'items_id');
+   $migration->addKey('glpi_items_softwarelicenses', [
+      'itemtype',
+      'items_id'
+   ], 'item');
+
+   $migration->changeField('glpi_configs', 'purge_computer_software_install',
```
Comment (cedric-anne) [inline install/update_94_95.php]: This is not a field, but a value.
Should be something like that:
```php
   $migration->addPostQuery(
      $DB->buildUpdate(
         'glpi_configs',
         ['name' => 'purge_item_software_install'],
         ['name' => 'purge_computer_software_install', 'context' => 'core']
      )
   );
```

Comment (trasher) [reply install/update_94_95.php]: There is a method in the Migration class to add/change configuration values (Migration::addConfig)

Comment (cconard96) [reply install/update_94_95.php]: I don't see a method for changing config names in Migration (or Config). The addConfig method only accepts name => value pairs and ignores existing configs.

Comment (trasher) [reply install/update_94_95.php]: I was indeed wrong, existing values are ignored there. addPostQuuery is the best option then ;)

**Comment by cedric-anne** [comment]
Deprecation warns make test failing as logs having level `>= Logger::NOTICE` are transformed into a `RuntimeException` when occuring inside automated tests.
Your backward compatibility tests are interesting, so I do not want them to be removed. @trasher How can we fix this ?

**Comment by trasher** [comment]
This cannot be fixed; deprecated methods cannot be tested; and honestly, I'm not sure it is so interesting.

Most of the tim, deprecated calls are replaced with anoher one, and if the new one is tested, it is not needed to test the old one.

**Comment by cconard96** [comment]
The only reason the compatibility code (mapping old fields to new ones) and tests were made was to keep existing scripts and code working without needing changed.

I'm not attached to the compatibility/test code though and will remove it if needed.

File: inc/api.class.php [hunk 3]
```diff
@@ -608,9 +608,7 @@ protected function getItem($itemtype, $id, $params = []) {
       }
 
       // retrieve computer softwares
-      if (isset($params['with_softwares'])
-          && $params['with_softwares']
-          && $itemtype == "Computer") {
+      if (isset($params['with_softwares']) && $params['with_softwares']) {
```
Comment (trasher) [inline inc/api.class.php]: As far as I've understood, `$itemtype` must be one of `$CFG_GLPI['software_types']`; this should be checked here

**Comment by orthagh** [comment]
> This feature is planned, but I haven't decided on a target version yet. I don't know what other agents/plugins inventory for Android/iOS, but I can certainly open another PR for Volumes (and maybe hard drive components) if you think they should be linked to Phones.

Yes please, plan this if you have time (and want to). Everybody (on our side) seems to agree for the direction pushed by this pr, so we need to finish the work for having smartphones in Assets > Phones.

**Comment by trasher** [comment]
There are conflicts, could you please rebase to get them fixed? Thanks

**Pull request merged**